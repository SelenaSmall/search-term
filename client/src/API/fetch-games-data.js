import axios from 'axios';

const fetchGamesData = async () => {
  const url = process.env.REACT_APP_API_BASE_URL || '';

  const response = await axios.get(`${url}/api/v1/games`);
  return response.data;
};

export default fetchGamesData;
