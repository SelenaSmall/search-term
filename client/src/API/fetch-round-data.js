import axios from 'axios';

const fetchRoundData = async (id) => {
  const url = process.env.REACT_APP_API_BASE_URL || '';

  const response = await axios.get(`${url}/api/v1/rounds/${id}`);
  return response.data;
};

export default fetchRoundData;
