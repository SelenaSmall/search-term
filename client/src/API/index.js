import axios from 'axios';

import fetchRoundData from './fetch-round-data';
import fetchGameData from './fetch-game-data';
import fetchGamesData from './fetch-games-data';
import newMatch from './new-match';

const joinMatch = async (playerId, matchId) => {
  const url = process.env.REACT_APP_API_BASE_URL || '';

  const response = await axios.put(`${url}/matches/${matchId}/join`, {
    player_id: playerId,
    match_id: matchId,
  });
  return response.data;
};

const fetchOrCreatePlayer = async (id) => {
  const url = process.env.REACT_APP_API_BASE_URL || '';

  const response = await axios.post(`${url}/players/${id}/first_or_create`, {
    playerId: id,
  });
  return response.data;
};

export default {
  fetchRoundData,
  fetchGameData,
  fetchGamesData,
  newMatch,
  joinMatch,
  fetchOrCreatePlayer,
};
