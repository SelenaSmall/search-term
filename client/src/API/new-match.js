import axios from 'axios';

const newMatch = async () => {
  const url = process.env.REACT_APP_API_BASE_URL || '';

  const response = await axios.post(`${url}/matches`, {
    match: { title: '' },
  });
  return response.data;
};

export default newMatch;
