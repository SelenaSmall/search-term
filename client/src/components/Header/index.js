import React from 'react';

export default ({ status, score, secondsRemaining}) => (
  <section id="top-region">
    <menu className="status-menu">
      <ul>
        <li className="status-menu-item status">
          <span className="status-menu-item-label">Status</span>
          <span className="status-menu-item-value">{status}</span>
        </li>
        <li className="status-menu-item score">
          <span className="status-menu-item-label">Score</span>
          <span className="status-menu-item-value">{score}</span>
        </li>
        <li className="status-menu-item timer">
          <span className="status-menu-item-label">Time</span>
          <span className="status-menu-item-value">{secondsRemaining}</span>
        </li>
      </ul>
    </menu>
  </section>
);
