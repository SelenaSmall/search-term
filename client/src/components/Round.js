import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { Row, Col } from 'reactstrap';
import API from '../API';
import './game.css';
import Header from './Header';

class Round extends Component {
  constructor(props) {
    super(props);
    const round = parseInt(props.match.params.round, 10);
    const gameName = props.match.params.gameName;
    this.state = {
      term: null,
      round,
      images: [],
      // status: 'IN-PROGRESS',
      status: 'LOADING ...', // TODO live fix
      secondsRemaining: 0,
      score: 0,
      interval: null,
      gameName,
      terms: [],
    };
    this.fetchGameData(gameName);
    const { gameStyle } = this.state;
    this.fetchRoundData(round, gameName, gameStyle);
    this.guessInputRef = React.createRef();
    this.nextButtonRef = React.createRef();
    this.handleGuess = this.handleGuess.bind(this);
    this.startTimer = this.startTimer.bind(this);
    this.tick = this.tick.bind(this);
    this.timeIsUp = this.timeIsUp.bind(this);
  }

  componentDidMount() {
    // this.guessInputRef.current.focus(); // TODO breaks tests
  }

  componentWillReceiveProps(props) {
    const { interval } = this.state;
    clearInterval(interval);
    const round = parseInt(props.match.params.round, 10);
    const gameName = props.match.params.gameName;
    this.setState({ round, secondsRemaining: 60, guess: '', gameName });
    this.fetchRoundData(round, gameName);
  }

  fetchRoundData(round, gameName, gameStyle) {
    API.fetchRoundData(round, gameName).then((data) => {
      this.setState(data);
      this.setState({ status: 'IN-PROGRESS' }); // TODO live fix
      this.startTimer();
      if (gameStyle === 'text') {
        this.guessInputRef.current.focus();
      }
    }).catch(() => {
      this.setState({ status: 'ERROR' });
    });
  }

  fetchGameData(gameId) {
    API.fetchGameData(gameId).then((data) => {
      this.setState({ maxRounds: data.rounds, gameStyle: data.game_style, secondsRemaining: data.seconds_per_round });
    });
  }

  startTimer() {
    const interval = setInterval(this.tick, 1000);
    this.setState({ interval });
  }

  tick() {
    const { secondsRemaining } = this.state;
    if (secondsRemaining <= 0) {
      this.timeIsUp();
    } else {
      (secondsRemaining);
      this.setState({ secondsRemaining: secondsRemaining - 1 })
    }
  }

  timeIsUp() {
    const { interval } = this.state;
    clearInterval(interval);
    this.setState({ status: 'TIME-IS-UP' });
  }

  handleGuess(event) {
    const { value } = event.target;
    const {
      term, interval, score, secondsRemaining, gameStyle,
    } = this.state;
    this.setState({ guess: value });
    const REGEX = new RegExp(term);
    if (REGEX.exec(value)) {
      this.setState({ status: 'WINNER' });
      clearInterval(interval);
      this.setState({ score: score + secondsRemaining });
      this.nextButtonRef.current && this.nextButtonRef.current.focus();
    } else {
      this.setState({ status: 'WRONG' });
      if (gameStyle === 'multi_choice') {
        clearInterval(interval);
        this.setState({ score: score + secondsRemaining });
        this.nextButtonRef.current && this.nextButtonRef.current.focus();
      }
    }
  }

  render() {
    const {
      status, score, secondsRemaining, round, maxRounds, images, guess, gameName, gameStyle, terms,
    } = this.state;

    const choices = terms.map( term => <li><button class="bg-blue hover:bg-blue-dark text-white font-bold py-2 px-4 rounded min-w-full mb-1" onClick={this.handleGuess} value={term} key={term}>{term}</button></li> );

    return (
      <div className="flex flex-column">
        <Header
          status={status}
          score={score}
          secondsRemaining={secondsRemaining}
        />

        <section id="middle-region">
          <h1 className="text-center">
Round
            <span className="round">{round}</span>
          </h1>

          <Row className="main-container-row">
            <Col sm="6">
              <div className="images-container">
                { status === 'LOADING ...'
                  ? <div className="background" />
                  : images.map(({ id, src }) => (
                    <div className="images-container-single" key={id}>
                      <img className="images-container-single-image" key={id} src={src} alt="nice try" />
                    </div>
                  ))}
              </div>
            </Col>

            <Col sm="6">
              <div className="input-container">
                <h5>Guess the search term</h5>
                {
                  gameStyle === 'text'
                    ? <textarea
                        ref={this.guessInputRef}
                        className="input-container-guess"
                        onChange={this.handleGuess}
                        value={guess}
                        disabled={status === 'WINNER'}
                      />
                    : <ul class="list-reset">{choices}</ul>
                }

                {
                  round >= maxRounds
                    ? <Link to={{ pathname: '/results', state: { score } }} className="input-container-next-round next-round game-button">Next</Link>
                    : <Link to={`/game/${gameName}/round/${round + 1}`} className="input-container-next-round next-round game-button">Next</Link>
                }
              </div>
            </Col>
          </Row>
        </section>
      </div>
    );
  }
}

export default Round;
