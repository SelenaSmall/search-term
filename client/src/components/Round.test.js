import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';

import Adapter from 'enzyme-adapter-react-16';
import Round from './Round';
import API from '../API';
import Header from './Header';

Enzyme.configure({ adapter: new Adapter() });

it('snapshot testing', () => {
  API.fetchRoundData = () => Promise.resolve();
  API.fetchGameData = () => Promise.resolve({ rounds: 1 });

  const output = shallow(<Round match={{ params: { round: 1 } }} />);
  expect(shallowToJson(output)).toMatchSnapshot();
});

it('calls api to get all images on construction', () => {
  const fetchRoundDataCalls = [];

  const roundDataPromise = Promise.resolve(
    {
      round: 1,
      images: [
        { src: 'image_1.jpg', id: 'abc-Xyz' },
        { src: 'image_1.jpg', id: 'abc-Xyz' },
      ],
      term: 'search term',
    },
  );
  API.fetchRoundData = (args) => {
    fetchRoundDataCalls.push(args);
    return roundDataPromise;
  };
  API.fetchGameData = () => Promise.resolve({ rounds: 1 });

  const output = shallow(<Round match={{ params: { round: 1 } }} />);
  roundDataPromise.then(() => {
    expect(shallowToJson(output)).toMatchSnapshot();
  });
  expect(fetchRoundDataCalls).toEqual([1]);
});

// TODO attempt to test active element
// https://stackoverflow.com/questions/37694900/testing-input-focus-in-enzyme/49448639#49448639
//   TypeError: Cannot read property 'focus' of null
//   console.log(this.guessInputRef)
//   { current: null }
//
// https://blog.logrocket.com/how-to-use-react-createref-ea014ad09dba
it.skip('focuses on the guess box by default', () => {
  API.fetchRoundData = () => Promise.resolve({});
  API.fetchGameData = () => Promise.resolve({ rounds: 1 });
  const wrapper = mount(
    <BrowserRouter>
      <Round match={{ params: { round: 1 } }} />
    </BrowserRouter>,
  );

  const focusedElement = document.activeElement;

  expect(wrapper.find('.input-container-guess').matchesElement(focusedElement)).toEqual(true);
});

it('calls api and handles an error', () => {
  const fetchRoundDataCalls = [];

  const expectedError = new Error('Fail');
  const roundDataPromise = Promise.reject(expectedError);

  API.fetchRoundData = (args) => {
    fetchRoundDataCalls.push(args);
    return roundDataPromise;
  };
  API.fetchGameData = () => Promise.resolve({ rounds: 1 });

  const output = shallow(<Round match={{ params: { round: 1 } }} />);
  expect(shallowToJson(output)).toMatchSnapshot();
  roundDataPromise.then(() => {
  }).catch(() => {
    expect(fetchRoundDataCalls).toEqual([1]);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});

it.skip('sets state to winner if guess is correct', () => {
  API.fetchGameData = () => Promise.resolve({ rounds: 1 });
  API.fetchRoundData = () => Promise.resolve({ term: 'ghost' });

  const output = shallow(<Round match={{ params: { round: 1 } }} />);
  expect(output.find('.status .status-menu-item-value').text()).toEqual('IN-PROGRESS');
  output.find('textarea.input-container-guess').simulate('change', { target: { value: 'ghost' } });
  expect(output.find('.status .status-menu-item-value').text()).toEqual('WINNER');
});

describe('status', () => {
  it('is set to LOADING ... by default', () => {
    const wrapper = shallow(<Round match={{ params: { round: 1 } }} />);
    expect(wrapper.find(Header).at(0).prop('status')).toBe('LOADING ...');
  });

  it('is set to IN-PROGRESS once the API call comes back', () => {
    expect.assertions(2);

    const roundDataPromise = Promise.resolve({ term: 'bjorn borg' });
    API.fetchRoundData = () => roundDataPromise;

    const wrapper = shallow(<Round match={{ params: { round: 1 } }} />);

    expect(wrapper.find(Header).at(0).prop('status')).toBe('LOADING ...');
    roundDataPromise.then(() => {
      expect(wrapper.find(Header).at(0).prop('status')).toBe('IN-PROGRESS');
    });
  });

  it.skip('is set to ERROR if the API call fails', () => {
    expect.assertions(2);

    const roundDataPromise = Promise.reject(new Error('oops'));
    API.fetchRoundData = () => roundDataPromise;

    const wrapper = shallow(<Round match={{ params: { round: 1 } }} />);

    expect(wrapper.find(Header).at(0).prop('status')).toBe('LOADING ...');
    roundDataPromise.catch(() => {
      expect(wrapper.find(Header).at(0).prop('status')).toBe('ERROR');
    });
  });

  it('is set to WINNER the right answer is guessed', () => {
    expect.assertions(3);

    const roundDataPromise = Promise.resolve({ term: 'bjorn borg' });
    API.fetchRoundData = () => roundDataPromise;

    const wrapper = shallow(<Round match={{ params: { round: 1 } }} />);

    expect(wrapper.find(Header).at(0).prop('status')).toBe('LOADING ...');
    roundDataPromise.then(() => {
      expect(wrapper.find(Header).at(0).prop('status')).toBe('IN-PROGRESS');
      wrapper.instance().handleGuess({target: { value: 'bjorn borg'}});
      expect(wrapper.find(Header).at(0).prop('status')).toBe('WINNER');
    });
  });

  it('is set to TIME-IS-UP once the time runs out', () => {
    expect.assertions(3);

    const roundDataPromise = Promise.resolve({ term: 'bjorn borg' });
    API.fetchRoundData = () => roundDataPromise;

    const wrapper = shallow(<Round match={{ params: { round: 1 } }} />);

    expect(wrapper.find(Header).at(0).prop('status')).toBe('LOADING ...');
    roundDataPromise.then(() => {
      expect(wrapper.find(Header).at(0).prop('status')).toBe('IN-PROGRESS');
      wrapper.instance().timeIsUp();
      expect(wrapper.find(Header).at(0).prop('status')).toBe('TIME-IS-UP');
    });
  });

  it('is set to WRONG if the wrong answer is guessed', () => {
    expect.assertions(3);

    const roundDataPromise = Promise.resolve({ term: 'bjorn borg' });
    API.fetchRoundData = () => roundDataPromise;

    const wrapper = shallow(<Round match={{ params: { round: 1 } }} />);

    expect(wrapper.find(Header).at(0).prop('status')).toBe('LOADING ...');
    roundDataPromise.then(() => {
      expect(wrapper.find(Header).at(0).prop('status')).toBe('IN-PROGRESS');
      wrapper.instance().handleGuess({target: { value: 'martina navratilova'}});
      expect(wrapper.find(Header).at(0).prop('status')).toBe('WRONG');
    });
  });
});
