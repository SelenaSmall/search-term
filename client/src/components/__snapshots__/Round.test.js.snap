// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls api to get all images on construction 1`] = `
<div>
  <section
    id="top-region"
  >
    <menu
      className="status-menu"
    >
      <ul>
        <li
          className="status-menu-item status"
        >
          <span
            className="status-menu-item-label"
          >
            Status
          </span>
          <span
            className="status-menu-item-value"
          >
            IN-PROGRESS
          </span>
        </li>
        <li
          className="status-menu-item score"
        >
          <span
            className="status-menu-item-label"
          >
            Score
          </span>
          <span
            className="status-menu-item-value"
          >
            0
          </span>
        </li>
        <li
          className="status-menu-item timer"
        >
          <span
            className="status-menu-item-label"
          >
            Time
          </span>
          <span
            className="status-menu-item-value"
          >
            10
          </span>
        </li>
      </ul>
    </menu>
  </section>
  <section
    id="middle-region"
  >
    <h1
      className="text-center"
    >
      Round
      <span
        className="round"
      >
        1
      </span>
    </h1>
    <Row
      className="main-container-row"
      tag="div"
    >
      <Col
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <div
          className="images-container"
        >
          <div
            className="images-container-single"
          >
            <img
              alt="nice try"
              className="images-container-single-image"
              key="abc-Xyz"
              src="image_1.jpg"
            />
          </div>
          <div
            className="images-container-single"
          >
            <img
              alt="nice try"
              className="images-container-single-image"
              key="abc-Xyz"
              src="image_1.jpg"
            />
          </div>
        </div>
      </Col>
      <Col
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <div
          className="input-container"
        >
          <h5>
            Guess the search term
          </h5>
          <textarea
            className="input-container-guess"
            onChange={[Function]}
          />
          <Link
            className="input-container-next-round next-round game-button"
            replace={false}
            to={
              Object {
                "pathname": "/results",
                "state": Object {
                  "score": 0,
                },
              }
            }
          >
            Next
          </Link>
        </div>
      </Col>
    </Row>
  </section>
  <section
    id="bottom-region"
  >
    <menu
      className="twitter-menu"
    >
      <ul>
        <li
          className="twitter-menu-item twitter-menu-item-left status"
        >
          <a
            href="http://twitter.com/intent/user?screen_name=selenasmall88"
          >
            <img
              alt="selena-small"
              className="twitter-menu-item-avatar"
              src="https://twitter.com/selenasmall88/profile_image?size=normal"
            />
            <span>
              @selenasmall88
            </span>
          </a>
        </li>
        <li
          className="twitter-menu-item status"
        >
          <a
            href="http://twitter.com/intent/user?screen_name=saramic"
          >
            <img
              alt="michael-milewski"
              className="twitter-menu-item-avatar"
              src="https://twitter.com/saramic/profile_image?size=normal"
            />
            <span>
              @saramic
            </span>
          </a>
        </li>
      </ul>
    </menu>
  </section>
</div>
`;

exports[`snapshot testing 1`] = `
<div>
  <section
    id="top-region"
  >
    <menu
      className="status-menu"
    >
      <ul>
        <li
          className="status-menu-item status"
        >
          <span
            className="status-menu-item-label"
          >
            Status
          </span>
          <span
            className="status-menu-item-value"
          >
            LOADING ...
          </span>
        </li>
        <li
          className="status-menu-item score"
        >
          <span
            className="status-menu-item-label"
          >
            Score
          </span>
          <span
            className="status-menu-item-value"
          >
            0
          </span>
        </li>
        <li
          className="status-menu-item timer"
        >
          <span
            className="status-menu-item-label"
          >
            Time
          </span>
          <span
            className="status-menu-item-value"
          >
            10
          </span>
        </li>
      </ul>
    </menu>
  </section>
  <section
    id="middle-region"
  >
    <h1
      className="text-center"
    >
      Round
      <span
        className="round"
      >
        1
      </span>
    </h1>
    <Row
      className="main-container-row"
      tag="div"
    >
      <Col
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <div
          className="images-container"
        />
      </Col>
      <Col
        sm="6"
        tag="div"
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
      >
        <div
          className="input-container"
        >
          <h5>
            Guess the search term
          </h5>
          <textarea
            className="input-container-guess"
            onChange={[Function]}
          />
          <Link
            className="input-container-next-round next-round game-button"
            replace={false}
            to="/round/2"
          >
            Next
          </Link>
        </div>
      </Col>
    </Row>
  </section>
  <section
    id="bottom-region"
  >
    <menu
      className="twitter-menu"
    >
      <ul>
        <li
          className="twitter-menu-item twitter-menu-item-left status"
        >
          <a
            href="http://twitter.com/intent/user?screen_name=selenasmall88"
          >
            <img
              alt="selena-small"
              className="twitter-menu-item-avatar"
              src="https://twitter.com/selenasmall88/profile_image?size=normal"
            />
            <span>
              @selenasmall88
            </span>
          </a>
        </li>
        <li
          className="twitter-menu-item status"
        >
          <a
            href="http://twitter.com/intent/user?screen_name=saramic"
          >
            <img
              alt="michael-milewski"
              className="twitter-menu-item-avatar"
              src="https://twitter.com/saramic/profile_image?size=normal"
            />
            <span>
              @saramic
            </span>
          </a>
        </li>
      </ul>
    </menu>
  </section>
</div>
`;
